src/shadowsocks_config.ts

```diff
@@ -338,3 +339,48 @@ export const SIP002_URI = {
     return `ss://${userInfo}@${uriHost}:${port.data}/${queryString}${hash}`;
   },
 };
+
+export interface ConfigFetchParams {
+  // URL endpoint to retrieve a Shadowsocks configuration.
+  readonly location: string;
+  // Server cerficate hash.
+  readonly certFingerprint?: string;
+  // HTTP method to use when accessing `url`.
+  readonly httpMethod?: string;
+}
+
+export const ONLINE_CONFIG_PROTOCOL = 'ssconf';
+
+// Parses access parameters to retrieve a Shadowsocks proxy config from an
+// online config URL. See: https://github.com/shadowsocks/shadowsocks-org/issues/89
+export function parseOnlineConfigUrl(url: string): ConfigFetchParams {
+  if (!url || !url.startsWith(`${ONLINE_CONFIG_PROTOCOL}:`)) {
+    throw new InvalidUri(`URI protocol must be "${ONLINE_CONFIG_PROTOCOL}"`);
+  }
+  // Replace the protocol "ssconf" with "https" to ensure correct results,
+  // otherwise some Safari versions fail to parse it.
+  const inputForUrlParser = url.replace(new RegExp(`^${ONLINE_CONFIG_PROTOCOL}`), 'https');
+  // The built-in URL parser throws as desired when given URIs with invalid syntax.
+  const urlParserResult = new URL(inputForUrlParser);
+  // Use ValidatedConfigFields subclasses (Host, Port, Tag) to throw on validation failure.
+  const uriFormattedHost = urlParserResult.hostname;
+  let host: Host;
+  try {
+    host = new Host(uriFormattedHost);
+  } catch (_) {
+    // Could be IPv6 host formatted with surrounding brackets, so try stripping first and last
+    // characters. If this throws, give up and let the exception propagate.
+    host = new Host(uriFormattedHost.substring(1, uriFormattedHost.length - 1));
+  }
+  // The default URL parser fails to recognize the default HTTPs port (443).
+  const port = new Port(urlParserResult.port || '443');
+  // Parse extra parameters from the tag, which has the URL search parameters format.
+  const tag = new Tag(decodeURIComponent(urlParserResult.hash.substring(1)));
```

Comment:
You are double decoding, which will break with the value has an encoded `&`, for example: 'https://dsfdsf.com#foo=a%26b'

Remove the decodeURIComponent call since URLSearchParams already decodes for you.

Also add a test for that case.

Reply:
Good catch. Done in fb81d34.